### PR TITLE
162/use is order unfillable

### LINKED ISSUE #162: Warning for orders that are unfillable
(price moved beyond slippage)
AFTER confirmation modal and price changes, maybe show warning in "Pending" order history

Confirmation modal will warn user when price changes before submitting, but once you send that order nothing will warn the user if the price changes after submitting and that it may not fill.

Could add a warning to the "pending" order notification in the toast or in the History Transactions comp

When cancellation is possible users can then leverage that option as to not have failing tx.

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Fixes #162 

Checking whether orders are unfillable

# To Test

1. On rinkeby (because it's easier to move the price of a fake token), pick GNO-WETH pair
2. Make sure you have about 5ETH worth of rinkeby GNO. During testing it was enough
3. Sell the equivalent of 2.5 WETH in GNO for WETH.
4. Right after, place a second order doing the same thing
5. Wait until one of the orders is matched
- [ ] The second order will no longer be "matchable", because the price should have moved against you (this depends on the pair liquidity, so YMMV)
- [ ] You should see a (very ugly) message under activity details stating that the price is out of range
screenshot_2021-08-06_16-19-43
- [ ] If you look at the console, you should see the pending order has a percentage greater than 1%, that represents the different between the first number (order price) to the second number (current price)
screenshot_2021-08-06_16-29-03

# Background

# Out of scope
- Make it look nice (Michel is working on that)
- Check against user's balance

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @alfetopito , works as you have described. 
However, I have a question: What should happen if a user sees 'Price out of range' order? Or what is supposed a user should do in this case? Wait till it expires or make him/her cancel it?
If we want to prompt him/her to cancel it, it would be nice to show a notification/modal  about out of range price.. Something like 'Your order's price is out of range. Would you like to cancel it?' or something like this... WDYT?

**Comment by alfetopito** [comment]
> Hey @alfetopito , works as you have described.
> However, I have a question: What should happen if a user sees 'Price out of range' order? Or what is supposed a user should do in this case? Wait till it expires or make him/her cancel it?
> If we want to prompt him/her to cancel it, it would be nice to show a notification/modal about out of range price.. Something like 'Your order's price is out of range. Would you like to cancel it?' or something like this... WDYT?

The UI/UX behaviour is being addressed by Michel in the activity details re-work and is out of the scope of this change.
The idea is to do like you said, allowing the user to see their order is no longer fillable.

This PR is only about detecting when that's the case

File: src/custom/state/orders/updaters/index.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export { EventUpdater } from './EventUpdater'
+export { CancelledOrdersUpdater } from './CancelledOrdersUpdater'
+export { UnfillableOrdersUpdater } from './UnfillableOrdersUpdater'
```
Comment (alfetopito) [inline src/custom/state/orders/updaters/index.ts:3]: Refactored `state/orders/updater.ts` into `state/orders/updaters/index.ts, EventUpdater.ts, CancelledOrdersUpdater.ts, UnfillatbleOrdersUpdater.ts and utils.ts`

**Comment by W3stside** [comment]
couple things:
1. there are now 3 updaters for orders and some do similar things. Are we not worried about further race conditions there?
2. `EventUpdater` seems an inappropriate name for the orders updating. Maybe make it more inline with the other names unless I'm missing sth there

File: src/custom/state/orders/consts.ts [hunk 2]
```diff
@@ -7,3 +8,6 @@ export const ContractDeploymentBlocks: Partial<Record<ChainId, number>> = {
 }
 
 export const OPERATOR_API_POLL_INTERVAL = 10000 // in ms
+export const PENDING_ORDERS_PRICE_CHECK_POLL_INTERVAL = 30000 // in ms
+
+export const OUT_OF_MARKET_PRICE_DELTA_PERCENTAGE = new Percent(1, 100) // 1/100 => 0.01 => 1%
```
Comment (W3stside) [inline src/custom/state/orders/consts.ts:13]: nit: `1%` as a comment is clear enough 😂

File: src/custom/state/orders/updaters/EventUpdater.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        (acc, { status, popupData }) => {
+          popupData && acc[status].push(popupData)
+          return acc
+        },
+        { fulfilled: [], expired: [], cancelled: [], unknown: [], pending: [] }
+      )
+
+      // Bach state update per group, if any
+
+      fulfilled.length > 0 &&
```
Comment (W3stside) [inline src/custom/state/orders/updaters/EventUpdater.ts:55]: this wont clash with the CancelledOrdersUpdater logic touching similar state?

**Review by W3stside — APPROVED** [review]
complicated, but works well and readable. nice job. maybe we make these badbois into a smaller pr next time but i guess this was rather inter-connected code.

File: src/custom/state/orders/consts.ts (hunk 2 above)
Comment (alfetopito) [reply src/custom/state/orders/consts.ts:13]: Well, I like to be extra clear ;)

Also, there are other places where I saw Percent(1, 10000) and I confess I have to think awhile to understand what it does.

File: src/custom/state/orders/updaters/EventUpdater.ts (hunk 3 above)
Comment (alfetopito) [reply src/custom/state/orders/updaters/EventUpdater.ts:55]: Well, not clash, but could have some race conditions, yes...

The (not very aptly named `EventUpdater`) checks for pending orders, and update their statuses, triggering toast notifications when they change to fulfilled, expired or cancelled.

The CancelledOrdersUpdater checks only for cancelled orders that might have been set to cancelled too soon.
It can happen that an order is considered `cancelled` but a solver submit a solution to it anyway.
And only for 5min after an order is cancelled.

The last (and new, introduced here) is the UnfillableOrdersUpdater, that checks the price of pending orders and updates a flag in the storage.

**Comment by alfetopito** [comment]
> couple things:
> 
>     1. there are now 3 updaters for orders and some do similar things. Are we not worried about further race conditions there?

We have 2 until now :)
Yes, there might be race conditions.
But I'm not sure what would be a better approach, since each updater watches a different set of orders and acts upon them in a different way.
Making it a single updater will create a big moster :scream: 

Any suggestions?
 
>     2. `EventUpdater` seems an inappropriate name for the orders updating. Maybe make it more inline with the other names unless I'm missing sth there

Agreed. Historical name, from when we actually listened for events to update the state. Will change the name.

**Comment by alfetopito** [comment]
> complicated, but works well and readable. nice job. maybe we make these badbois into a smaller pr next time but i guess this was rather inter-connected code.

I considered having the refactor (split single updater into 3 files) in a different but then, naaaaah.

Next time, promise :crossed_fingers:

File: src/custom/state/orders/consts.ts (hunk 2 above)
Comment (anxolin) [reply src/custom/state/orders/consts.ts:13]: Hehe, I like the extra clarity :)

**Comment by anxolin** [comment]
> Yes, there might be race conditions.

Why are we saying we can have race conditions? Can we give a sequence that would leave a bad state? 
Afaik, if they touch different fields of the order, since JS is executed with just one thread, the updaters should not conflict and break the state afait

File: src/custom/state/orders/hooks.ts [hunk 4]
```diff
@@ -58,6 +59,7 @@ type CancelOrdersBatchParams = UpdateOrdersBatchParams
 interface UpdateLastCheckedBlockParams extends ClearOrdersParams {
   lastCheckedBlock: number
 }
+type SetIsOrderUnfillableParams = { id: OrderID; chainId: ChainId; isUnfillable: boolean }
```
Comment (anxolin) [inline src/custom/state/orders/hooks.ts:62]: Is this a duplicate from `SetIsOrderUnfillableParams` in the `actions`?

File: src/custom/state/orders/updaters/CancelledOrdersUpdater.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+ *
+ * Whenever an order that was cancelled but has since been fulfilled, trigger a state update
+ * and a popup notification, changing the status from cancelled to fulfilled.
+ *
+ * It's supposed to fix race conditions between the api accepting a cancellation while a solution was already
+ * submitted to the network by a solver.
+ * Due to the network's nature, we can't tell whether an order has been really cancelled, so we prefer to wait a short
+ * period and say it's cancelled even though in some cases it might actually be filled.
+ */
+export function CancelledOrdersUpdater(): null {
```
Comment (anxolin) [inline src/custom/state/orders/updaters/CancelledOrdersUpdater.ts:29]: Ideally, in one previous PR you moved these other PRs code so it doesn't look like is part of this one. 
I'm not reviewing this code cause im assuming theres no changes in the `Updaters` (only in unfillable)

File: src/custom/state/orders/consts.ts [hunk 6]
```diff
@@ -7,3 +8,6 @@ export const ContractDeploymentBlocks: Partial<Record<ChainId, number>> = {
 }
 
 export const OPERATOR_API_POLL_INTERVAL = 10000 // in ms
+export const PENDING_ORDERS_PRICE_CHECK_POLL_INTERVAL = 30000 // in ms
```
Comment (anxolin) [inline src/custom/state/orders/consts.ts:11]: Is it to bad to do it every 10s? 

There's a bunch of stuff (like this one) that I would love to change it to logic we trigger on every new block instead of at regular intervals. 

10s is just a rudimentary way to make sure we update on every block.

File: src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+ * Out of the market means the price difference between original and current to be positive
+ * and greater than OUT_OF_MARKET_PRICE_DELTA_PERCENTAGE.
+ * Negative difference is good for the user.
+ * We allow for it to be up to OUT_OF_MARKET_PRICE_DELTA_PERCENTAGE worse to account for
+ * small price changes
+ *
+ * @param order
+ * @param price
+ */
+function isOrderUnfillable(order: Order, price: Required<PriceInformation>): boolean {
```
Comment (anxolin) [inline src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts:62]: I like how you implemented all, you break things in small functions. 

Would it be worth it to do in another PR some test for this nice function, since it's a pure function and key for the logic in question?

File: src/custom/state/orders/consts.ts (hunk 6 above)
Comment (anxolin) [reply src/custom/state/orders/consts.ts:11]: Ignore me if 10s will risk hitting time limit of some sort. I'm assuming they allow several per second, if it's not the case 30s as u suggest is good enough

**Review by anxolin — APPROVED** [review]
Great work! The code was easy to read and follow. 

This together with @biocom side menu, it's going to be great.

File: src/custom/state/orders/updaters/CancelledOrdersUpdater.ts (hunk 5 above)
Comment (alfetopito) [reply src/custom/state/orders/updaters/CancelledOrdersUpdater.ts:29]: I know, I know https://github.com/gnosis/cowswap/pull/1201#issuecomment-896372327

Was lazy, will not happen again, I learned my lesson, etc, etc

File: src/custom/state/orders/consts.ts (hunk 6 above)
Comment (alfetopito) [reply src/custom/state/orders/consts.ts:11]: I did not face any rate limiting issues.
I'm just being conservative in having stuff running on the brackground updating too often, to avoid making the page slow(er than usual) and overloading the query endpoints with too many requests.

I can change it to 10s if you think that's fine.

File: src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts (hunk 7 above)
Comment (alfetopito) [reply src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts:62]: Yep, will do.

File: src/custom/state/orders/consts.ts (hunk 6 above)
Comment (anxolin) [reply src/custom/state/orders/consts.ts:11]: It's alright. Make sense

File: src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts (hunk 7 above)
Comment (alfetopito) [reply src/custom/state/orders/updaters/UnfillableOrdersUpdater.ts:62]: https://github.com/gnosis/cowswap/pull/1260
